fastlane/lib/fastlane/actions/artifactory.rb

```diff
@@ -11,33 +11,44 @@ def self.run(params)
 
         require 'artifactory'
         file_path = File.absolute_path(params[:file])
-        if File.exist?(file_path)
-          client = connect_to_artifactory(params)
-          artifact = Artifactory::Resource::Artifact.new
-          artifact.client = client
-          artifact.local_path = file_path
-          artifact.checksums = {
-              "sha1" => Digest::SHA1.file(file_path),
-              "md5" => Digest::MD5.file(file_path)
-          }
-          UI.message("Uploading file: #{artifact.local_path} ...")
-          upload = artifact.upload(params[:repo], params[:repo_path], params[:properties])
 
-          Actions.lane_context[SharedValues::ARTIFACTORY_DOWNLOAD_URL] = upload.uri
-          Actions.lane_context[SharedValues::ARTIFACTORY_DOWNLOAD_SIZE] = upload.size
+        if (params[:username] && params[:password]) || params[:api_key]
+          if File.exist?(file_path)
+            client = connect_to_artifactory(params)
+            artifact = Artifactory::Resource::Artifact.new
+            artifact.client = client
+            artifact.local_path = file_path
+            artifact.checksums = {
+                "sha1" => Digest::SHA1.file(file_path),
+                "md5" => Digest::MD5.file(file_path)
+            }
+            UI.message("Uploading file: #{artifact.local_path} ...")
+            upload = artifact.upload(params[:repo], params[:repo_path], params[:properties])
 
-          UI.message("Uploaded Artifact:")
-          UI.message("Repo: #{upload.repo}")
-          UI.message("URI: #{upload.uri}")
-          UI.message("Size: #{upload.size}")
-          UI.message("SHA1: #{upload.sha1}")
+            Actions.lane_context[SharedValues::ARTIFACTORY_DOWNLOAD_URL] = upload.uri
+            Actions.lane_context[SharedValues::ARTIFACTORY_DOWNLOAD_SIZE] = upload.size
+
+            UI.message("Uploaded Artifact:")
+            UI.message("Repo: #{upload.repo}")
+            UI.message("URI: #{upload.uri}")
+            UI.message("Size: #{upload.size}")
+            UI.message("SHA1: #{upload.sha1}")
+          else
+            UI.message("File not found: '#{file_path}'")
+          end
         else
-          UI.message("File not found: '#{file_path}'")
+          if params[:username] && !params[:password]
+            UI.user_error!("Connect to artifactory using 'username' require setting a 'password'")
+          elsif !params[:username] && params[:password]
+            UI.user_error!("Connect to artifactory using 'password' require setting a 'password'")
+          else
+            UI.user_error!("You can either connect to artifactory using 'username' and 'password' or 'api_key'")
+          end
```

Comment:
I think it's more readable to validate these inputs on the top of this function using ruby one-liners, instead of adding more indentation with a `if/else` statement 🙇 
It'd look something like this:

```
UI.user_error!("Cannot connect to Artifactory - 'username' was provided but it's missing 'password'") if params[:username] && !params[:password]
UI.user_error!("Cannot connect to Artifactory - 'password' was provided but it's missing 'username'") if !params[:username] && params[:password]
UI.user_error!("Cannot connect to Artifactory - either 'api_key', or 'username' and 'password' must be provided") if !params[:api_key] && !params[:username]

if File.exist?(file_path) 
  …
else
  …
end
```

(also notice that I improved the copy of the error messages 😃